First cut changes for target traverser.

nitin-deamon opened the pull request.

nitin-deamon committed:
First cut changes for target traverser.

nitin-deamon committed:
Few fixes for target traverser.

- Added to get statx call in common.
- Added MemAvailable call in common.

This is code not compilable as it depend on someof struct defined in azcopy latest dev.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -39,13 +45,84 @@ type syncDestinationComparator struct {
 	// storing the source objects
 	sourceIndex *objectIndexer
 
+	sourceFolderIndex *folderIndexer
```
remove sourceIndex and add the following comment

        //
        // This stores the source files and their attributes as scanned by the source traverser.
        // Source traverser populates it using folderIndexer.store() while target traverser consumes
        // it using syncDestinationComparator.processIfNecessary().
        // Note that source traverser refers to this using syncEnumerator.objectIndexer so this and
        // syncEnumerator.objectIndexer should refer to the same folderIndexer.
        //

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -39,13 +45,84 @@ type syncDestinationComparator struct {
+	// Change file detection mode. Valid Enums will be TargetCompare, Ctime, CtimeMtime.
+	// TargetCompare - Default sync comparsion where target enumerated for each file. It's least optimized, but gurantee no data loss.
+	// Ctime - CTime used to detect change files/folder. Should be used where mtime not reliable.
+	// CtimeMtime - Both CTime      and MTime used to detect changed files/folder. It's most efficient in all of the cfdModes.
```
The comment in the doc is more detailed, we can use that directly.


	//
	// Change file detection mode.
	// This controls how target traverser decides whether a file has changed (and hence needs to be sync'ed
	// to the target) by looking at the file properties stored in the sourceFolderIndex.
	// Valid Enums will be TargetCompare, Ctime, CtimeMtime.
	//
	// TargetCompare - This is the most generic and the slowest of all. It enumerates all target directories
	//                 and compares the children at source and target to find out if an object has changed.
	//
	// CtimeMtime    - Compare source file’s mtime/ctime (Unix/NFS) or LastWriteTime/ChangeTime (Windows/SMB)
	//                 with LastSyncTime for detecting changed files. If mtime/LastWriteTime > LastSyncTime then
	//                 it means both data and metadata have changed else if ctime/ChangeTime > LastSyncTime then
	//                 only metadata has changed. For detecting changed directories this uses ctime/ChangeTime
	//                 (not mtime/LastWriteTime) of the directory, changed directories will be enumerated in the
	//                 target and every object will be compared with the source to find changes. This is needed
	//                 to cover the case of directory renames where a completely different directory in the source
	//                 can have the same name as a target directory and checking only mtime/LastWriteTime of children
	//                 is not safe since rename of a directory won’t affect mtime of its children. If a directory
	//                 has not changed then it’ll compare mtime/LastWriteTime and ctime/ChangeTime of its children
	//                 with LastSyncTime for finding data/metadata changes, thus it enumerates only target directories
	//                 that have changed.
	//                 This is the most efficient of all and should be used when we safely can use it, i.e., we know
	//                 that source updates ctime/mtime correctly.
	//
	// Ctime         - If we don’t want to depend on mtime/LastWriteTime (since it can be changed by applications)
	//                 but we still don’t want to lose the advantage of ctime/ChangeTime which is much more dependable,
	//                 we can use this mode. In this mode we use only ctime/ChangeTime to detect if there’s a change
	//                 to a file/directory and if there’s a change, to find the exact change (data/metadata/both) we
	//                 compare the file with the target. This has the advantage that we use the more robust
	//                 ctime/ChangeTime and only for files/directories which could have potentially changed we have
	//                 to compare with the target. This results in a much smaller set of files to enumerate at the
	//                 target.
	//
	// So, in the order of preference we have,
	//
	// CtimeMtime -> Ctime -> TargetCompare.
	//

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -39,13 +45,84 @@ type syncDestinationComparator struct {
+	// This is used as a lower bound to find out files/dirs changed since last sync.
+	// Depending on CFDMode this will be compared with source files' ctime/mtime or the target files' ctime/mtime.
+	//
+	lastSync time.Time
```
nit, let's call this lastSyncTime which is more explicit and aligns with the doc too.
also slight update to the comment

	//
	// Time of last sync, used by the sync process.
	// A file/directory having ctime/mtime greater than lastSyncTime is considered "changed", though the exact
	// interpretation depends on the CFDMode used and other variables.
	// Depending on CFDMode this will be compared with source files' ctime/mtime and/or target files' ctime/mtime.
	//

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -39,13 +45,84 @@ type syncDestinationComparator struct {
+	//
+	lastSync time.Time
+
+	// In case of metadata change only, shall we transfer whole file or only metadata. This flag governs that.
```
again, description from doc can be directly picked

	//
	// Sync only file metadata if only metadata has changed and not the file content, else for changed files both
	// file data and metadata are sync’ed. The latter could be expensive especially for cases where user updates
	// some file metadata (owner/mode/atime/mtime/etc) recursively. How we find out whether only metadata has
	// changed or only data has changed, or both have changed depends on the CFDMode that controls how changed
	// files are detected.
	//

and for TargetCtimeSkew

	//
	// This is the number of seconds that the Target’s clock is ahead of the Source’s clock. This will be subtracted
	// from Target’s ctime value before comparing with the Source’s ctime value when checking for “newness”. This will
	// be set *only* for targets which do not allow setting ctime and instead set ctime to the “now” value when a file
	// operation is executed. For targets where we set the ctime value we can safely compare for equality with the
	// source ctime value since we would have set it to source ctime value when the target object was sync’ed, for
	// such targets TargetCtimeSkew is set to 0.
	//
	// Note: If we choose a larger skew value, we might wrongly consider some object as needing sync, but that’s
	//       harmless as opposes to choosing a smaller skew and missing out syncing some object that has changed.
	//

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -39,13 +45,84 @@ type syncDestinationComparator struct {
+//
+// Note: This SHOULD NOT be called for children of "changed" directories, since for changed directories we cannot safely
+// 	     check for changed files purely by doing time based comparison. Use HasFileChangedSinceLastSyncUsingTargetCompare()
+// 		 for children of changed directories.
```
better to use the same comment as in doc

    //
    // Given a file and the corresponding scanned source object, find out if
    // we need to copy data+metadata, only metadata, or nothing.
    // This is called by TargetTraverser.
    // It honours various sync qualifiers to make the decision, f.e., if sync
    // qualifiers allow ctime/mtime to be used for CFD it may not need to
    // query file attributes from target.
    //  
    // Note: Caller will use the returned information to decide whether
    //       to copy the storedObject to target and whether to copy only metadata,
    //       or both metadata+data.
    //
    // Note: This SHOULD NOT be called for children of "changed" directories, since
    //       for changed directories we cannot safely check for changed files purely
    //       by doing local-time based comparison.
    //       Use HasFileChangedSinceLastSyncUsingTargetCompare() for children of
    //       changed directories.
    //       This means it will NEVER BE CALLED for CFDMode==TargetCompare, since for
    //       that HasDirectoryChangedSinceLastSync() always returns true, i.e., all
    //       directories are treated as “changed”.
    //
    // Return: (dataChanged, metadataChanged)
    //
    // Note: Since data change usually causes metadata change too (LMT is updated at the
    //       least), caller should check dataChanged first and if that is true, sync
    //       both data+metadata, if dataChanged is not true then it should check
    //       metadataChanged and if that is true, sync only metadata, else sync nothing.
    //

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -39,13 +45,84 @@ type syncDestinationComparator struct {
+// Note: This SHOULD NOT be called for children of "changed" directories, since for changed directories we cannot safely
+// 	     check for changed files purely by doing time based comparison. Use HasFileChangedSinceLastSyncUsingTargetCompare()
+// 		 for children of changed directories.
+func (f *syncDestinationComparator) HasFileChangedSinceLastSyncUsingLocalChecks(so StoredObject, filePath string) (dataChange bool, metadataChange bool) {
```
let's go over it and discuss. I want to make sure this exactly matches the doc, and if needed we update the doc.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +131,213 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+	// If mtime or size of target file is different from source file it means the file data (and metadata) has changed,
+	// else only metadata has changed.
+	//
+	if to.size != so.size || so.lastModifiedTime.After(to.lastModifiedTime) {
```
We cannot do the so.lastModifiedTime.After(to.lastModifiedTime) comparison.
It's important that we do a != comparison.
Again, pls don't deviate from the doc unless we have a good reason and in that case we need to update the doc.
I see there are some more deviations, let's discuss.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +131,213 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+			//       1. End Marker came for folder which is not present on target(only happen in case of root folder and first copy).
+			//			In that case folderPresent will be true. Otherwise this folder will be created by its parent.
+			//       2. End Marker came for folder present on target. In that case we have only newly files.
+
```
actually it's more readable to refactor the code as in the doc.
See FinalizeTargetDirectory  and the FinalizeAll flag.
The FinalizeAll flag should be set in the destinationObject depending on whether the target directory is being finalized after enumeration or w/o enumeration.
Let's discuss this.

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -77,7 +77,10 @@ type blobTraverser struct {
 
 	// cfdMode is change file detection mode. How to detect the file/folder changed.
 	// Changed files can be detected on basisi of ctime, ctimemtime , archiveBit or none.
```
replace L79 with

// See syncDestinationComparator.CFDMode for details.

same for L82

linuxsmiths replied:
Like in other places, do change the comments to refer to the actual place where we describe these

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,13 +239,83 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+//       This increases sync efficiency.
+//
+// Note: If we discover that certain sources cannot be safely trusted for ctime update we can change this to return True for them,
+//       thus falling back on the more rigorous target<->source comparison.
```
I slightly updated the comment in the doc. 


	//
	// Given a directory find out if it has changed since the last sync.
	// A “changed” directory could mean one or more of the following:
	// 1. One or more new files/subdirs created inside the directory.
	// 2. One or more files/subdirs deleted.
	// 3. Directory is renamed.
	//
	// It honours CFDMode to make the decision, f.e., if CFDMode allows ctime/mtime
	// to be used for CFD it may not need to query attributes from target.
	//
	// If it returns True, TargetTraverser will enumerate the directory and
	// compare each enumerated object with the source scanned objects in
	// ObjectIndexer[] to find out if the object needs to be sync'ed.
	//
	// For CFDModes that allow ctime for CFD we can avoid enumerating target
	// dir if we know directory has not changed. This increases sync efficiency.
	//
	// Note: If we discover that certain sources cannot be safely trusted for ctime
	//       update we can change this to return True for them, thus falling back
	//       on the more rigorous target<->source comparison.
	//

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,13 +239,83 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+// Note: If we discover that certain sources cannot be safely trusted for ctime update we can change this to return True for them,
+//       thus falling back on the more rigorous target<->source comparison.
+func (t *blobTraverser) HasDirectoryChangedSinceLastSync(so StoredObject, containerURL azblob.ContainerURL) bool {
+	// CFDMode is to enumerate Target and compare. So we return true in that case.
+	// This will lead to enumeration of target.
```
the comments in the doc are more appropriate, pls use them as-is.
f.e. here the following comment is apt

	        //
	        // Force enumeration for TargetCompare mode.
	        // For other CFDModes we enumerate a directory iff it has changed
	        // since last sync.
	        //

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,13 +239,83 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+//
+// Note: If we discover that certain sources cannot be safely trusted for ctime update we can change this to return True for them,
+//       thus falling back on the more rigorous target<->source comparison.
+func (t *blobTraverser) HasDirectoryChangedSinceLastSync(so StoredObject, containerURL azblob.ContainerURL) bool {
```
let's go over it and discuss

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,13 +239,83 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+				currentDirPath = currentDirPath + "/"
+			}
+
+			// Check the directory if we need enumeration at target side or not.
+			if !t.HasDirectoryChangedSinceLastSync(so, containerURL) {
+				goto FinalizeDirectory
+			}
```
we should use the same FinalizeAll boolean flag as in the doc and convey it using the storedObject enqueued at L429.
Let's discuss this.

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -466,7 +560,7 @@ func (t *blobTraverser) serialList(containerURL azblob.ContainerURL, containerNa
 func newBlobTraverser(rawURL *url.URL, p pipeline.Pipeline, ctx context.Context, recursive, includeDirectoryStubs bool, incrementEnumerationCounter enumerationCounterFunc,
 	s2sPreserveSourceTags bool, cpkOptions common.CpkOptions, indexerMap *folderIndexer, tqueue chan interface{}, isSource bool, isSync bool, maxObjectIndexerSizeInGB uint32,
-	lastSyncTime time.Time, cfdMode CFDModeFlags) (t *blobTraverser) {
+	lastSyncTime time.Time, cfdMode common.CFDMode, metaDataSyncOnly bool) (t *blobTraverser) {
```
metaDataSyncOnly  -> metaDataOnlySync

linuxsmiths commented on cmd/sync.go:
```diff
@@ -99,6 +99,30 @@ type rawSyncCmdArgs struct {
 	//
 	maxObjectIndexerMapSizeInGB string
+
+	// Change file detection flag. Valid Flag are TargetCompare, Ctime, CtimeMtime.
+	// TargetCompare - Default sync comparsion where target enumerated for each file. It's least optimized, but gurantee no data loss.
+	// Ctime - CTime used to detect change files/folder. Should be used where mtime not reliable.
+	// CtimeMtime - Both CTime      and MTime used to detect changed files/folder. It's most efficient in all of the cfdModes.
+	cfdMode string
+
+	// In case of metadata change only, shall we transfer whole file or only metadata. This flag governs that.
+	metaDataOnlySync bool
+
```
I have commented to add comments in syncDestinationComparator.
Instead, let's add all the comments here and refer all other places to here, instead of repeating the same comments in all places

nitin-deamon replied:
I added the comment in cookedSyncArgs and other places will refer it. The reason for adding those comments in cooked, they are more relevant there. Other places just refer them as it is, as they are passed without any modification. Some comments may be not relevant as lastSyncTime in rawSyncArgs. It tells more about format of lastSyncTime, which relevantin rawSyncArgs.

linuxsmiths commented on cmd/sync.go:
```diff
@@ -115,6 +139,19 @@ func (raw *rawSyncCmdArgs) parsePatterns(pattern string) (cookedPatterns []strin
+		return common.CFDModeFlags.TargetCompare()
+	} else if cfdMode == strings.ToLower(common.CFDModeFlags.Ctime().String()) {
+		return common.CFDModeFlags.Ctime()
+	} else if cfdMode == strings.ToLower(common.CFDModeFlags.CtimeMtime().String()) && raw.lastSyncTime != "" {
```
instead of doing the lastSyncTime check here, let the caller do it. It can log an appropriate message/warning too.
Also, lastSyncTime should be valid for both CtimeMtime and Ctime cfd modes, not just CtimeMtime

nitin-deamon replied:
sorry my bad, i don't want to check at all if time is nil or not. Reason for that customer want to try sync as first copy only then in that case lastSyncTime should be zero. I will add the comment in argument side for cfdMode.

linuxsmiths commented on cmd/sync.go:
```diff
@@ -461,7 +512,14 @@ type cookedSyncCmdArgs struct {
 	dryrunMode bool
 
-	cfdMode CFDModeFlags
+	// Change file detection Mode. Valid Enums are TargetCompare, Ctime, CtimeMtime.
+	// TargetCompare - Default sync comparsion where target enumerated for each file. It's least optimized, but gurantee no data loss.
+	// Ctime - CTime used to detect change files/folder. Should be used where mtime not reliable.
+	// CtimeMtime - Both CTime      and MTime used to detect changed files/folder. It's most efficient in all of the cfdModes.
+	cfdMode common.CFDMode
+
+	// In case of metadata change only, shall we transfer whole file or only metadata. This flag governs that.
+	metaDataOnlySync bool
```
again, refer to the comments in rawSyncCmdArgs.
Or do you think we should add comments to cookedSyncCmdArgs  and refer all other places to here?
let's discuss.

nitin-deamon replied:
cookedSyncCmdArgs will be better.

linuxsmiths commented on cmd/syncIndexer.go:
```diff
@@ -177,6 +178,63 @@ func (i *folderIndexer) store(storedObject StoredObject) (err error) {
 	return
 }
 
+// filesChangedInDirectory returns true in case any file in folder have ctime more than lastSyncTime.
```
enhance comment

 	//
	// This MUST be called only with CFDMode==Ctime and MetadataOnlySync==true.
	// This forces caller (target traverser) to use the more efficient ListDir for
	// getting file properties in bulk instead of querying individual file properties.
	//

linuxsmiths reviewed (commented):
Still not done reviewing completely. Pls look at these comments.

nitin-deamon committed:
Addressed the review comments -1

nitin-deamon committed:
Merge branch 'personal/nitin/target-traverser' of https://github.com/nitin-deamon/azure-storage-azcopy into personal/nitin/target-traverser

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,90 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+			panic("Directory entry is not string")
+		}
+
+		FinalizeAll := false
```
nit, comment above this

	        //
	        // Flag to be passed in FinalizeDirectory to indicate to the receiver (processIfNecessary()) if it
	        // should copy all the files or call HasFileChangedSinceLastSyncUsingLocalChecks() to find out files
	        // that need to be copied.
	        //

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,90 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+			return false
+		}
+	} else {
+		// Lets try to get the properties for directory.
+		blobURL := containerURL.NewBlobURL(strings.TrimSuffix(so.relativePath, common.AZCOPY_PATH_SEPARATOR_STRING))
+		resp, err := blobURL.GetProperties(context.TODO(), azblob.BlobAccessConditions{}, azblob.ClientProvidedKeyOptions{})
+		if err != nil {
+			return true
+		}
+		// Note: ContentLength gives the size of the blob in bytes. For a page blob, this header returns the value of the x-ms-blob-content-length header that is stored with the blob.
+		if so.lastModifiedTime.After(resp.LastModified()) || so.size != resp.ContentLength() {
+			return true
+		} else {
+			return false
+		}
```
we can get rid of this and add a 
panic(fmt.Sprintf("Unsupported CFDMode: %d", t.cfdMode));

then we can also get rid of containerURL argument.

Note that flat blobs will not store directory attributes, so blobURL.GetProperties() may not work.

nitin-deamon replied:
We should update the document in that case.

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,90 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+		return true
+	}
+
+	if t.cfdMode == common.CFDModeFlags.CtimeMtime() {
```
nit, add comment above this

	    //
	    // If CFDMode allows using ctime, compare directory ctime with LastSyncTime.
	    // Note that directory ctime will change if a new object is created inside the
	    // directory or an existing object is deleted or the directory is renamed.
	    //

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,90 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+		} else if t.metaDataOnlySync && t.indexerMap.filesChangedInDirectory(so.relativePath, t.lastSyncTime) {
+			// Listing of blob gives properties and each list api gives 5K blobs in single rest call. So if we know any file changed on source,
+			// We should list the target. It's optimization instead of getting properties for changed files one by one get properties of all blob.
+			// As in normal scenario, there will be less than 2K files in a folder. It make sense to call list api, if any file changed on source.
```
nit, we should also tell what we are doing here

			//
			// If metaDataOnlySync is true and a file has "ctime > LastSyncTime" and CFDMode does not allow us
			// to use mtime for checking if the file's data+metadata or only metadata has changed, then we need
			// to compare the file's source attributes with target attributes. Since fetching attributes for
			// individual target file may be expensive for some targets (f.e. Blob), so it would be better to
			// enumerate the target parent dir which will be cheaper due to ListDir returning many files and their
			// attributes in a single call. In normal scenario, there will be less than 2K files in a folder and
			// all 2K files along with their attributes can be retrieved in a single ListDir call.
                        // If not even one file changed in a directory we don't need to compare attributes for any file
                        // and hence we don't fetch the attributes.
	                //

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,90 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+				currentDirPath = currentDirPath + "/"
+			}
+
+			// Check the directory if we need enumeration at target side or not.
```
following comment is more helpful

	        //
	        // If source directory has not changed since last sync, then we don't really need to
	        // enumerate the target. SourceTraverser would have enumerated this directory and added
	        // all the children in ObjectIndexer map. We just need to go over these objects and find
	        // out which of these need to be sync'ed to target and sync them appropriately (data+metadata
	        // or only metadata).
	        //
	        // If directory has changed then there could be some files deleted and to find them we need
	        // to enumerate the target directory and compare. Also, if directory is renamed then also it'll
	        // be considered changed. Note that a renamed directory needs to be fully enumerated at the
	        // target as even files with same names as in the target could be entirely different files.
	        // This forces us to enumerate the target directory if the source directory is seen to have
	        // changed, since we don’t know if it was renamed, in which case we must enumerate the target
	        // directory.
	        //

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,90 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+		}
+
 		for marker := (azblob.Marker{}); marker.NotDone(); {
+			FinalizeAll = true
```
take this before the for loop and add this comment

	        //
	        // Now that we are going to enumerate the target dir, we will process all files present in
	        // the target, and when FinalizeDirectory is called only those files will be left in the ObjectIndexer
	        // map which are newly created in source. We need to blindly copy all of these to the
	        // target. Set FinalizeAll to true to convey this to the receiver (processIfNecessary()).
	        //

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,90 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
 
+		// This code for sync operation and when its target traverser.
+		if t.isSync && !t.isSource {
+			so := t.indexerMap.getStoredObject(currentDirPath)
```
very first thing, can we add

  //
  // sync target traverse should only enumerate direct children of a directory.
  //
 ```
 if t.recursive {
      panic(recursive enumeration flag set for sync target traverser!")
  }
```

nitin-deamon replied:
lets discuss as this flag cause source to enumerate recursively.

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -259,7 +338,9 @@ func (t *blobTraverser) parallelList(containerURL azblob.ContainerURL, container
 						resp, err := fblobURL.GetProperties(t.ctx, azblob.BlobAccessConditions{}, azblob.ClientProvidedKeyOptions{})
+						// TODO: Need to debug why its giving error here.
 						if err == nil {
+							extendedProp, _ := common.ReadStatFromMetadata(resp.NewMetadata(), resp.ContentLength())
```
I could not find the defintion of common.ReadStatFromMetadata()

also, why did you have to change code inside t.recursive(). We should not be running this code, right?

nitin-deamon replied:
ReadStatFromMetadata() defined in latest azcopy code adele.

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,90 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+			// source traverser (local traverser) truncate all the forward slash in storedObject.
+			// Without forward slash list blob not list blobs under this directory.
+			if currentDirPath != "" {
+				currentDirPath = currentDirPath + "/"
```
just to make sure we don't cause an extra "/" to be added which may not go down well later, let's do this

```
if currentDirPath[len(currentDirPath)-1]) != '/' {
    currentDirPath = currentDirPath + "/"
}
```

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -334,6 +417,24 @@ func (t *blobTraverser) parallelList(containerURL azblob.ContainerURL, container
+			storedObject := StoredObject{
+				name:              getObjectNameOnly(strings.TrimSuffix(currentDirPath, common.AZCOPY_PATH_SEPARATOR_STRING)),
+				relativePath:      strings.TrimSuffix(currentDirPath, common.AZCOPY_PATH_SEPARATOR_STRING),
+				entityType:        common.EEntityType.File(), // folder stubs are treated like files in in the serial lister as well
```
the comment "// folder stubs..." is not appropriate, since this is a special storedObject  that we use for marking end of directory and how we treat it is completely on us, so no need to compare with serial lister.

Also, I don't see why we should set it to File(). We should set it to Folder() since it is a folder.

nitin-deamon replied:
Yes it should be folder, previously we are setting isVirtualFolder = true for stubdir and endMarker. But in recent azcopy code i see for stub its set to folder. I changed for stub, but forgot to do it here.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -36,16 +42,155 @@ type syncDestinationComparator struct {
+func (f *syncDestinationComparator) HasFileChangedSinceLastSyncUsingLocalChecks(so StoredObject, filePath string) (dataChange bool, metadataChange bool) {
+	// CFDMode==TargetCompare always treats target directories as “changed”, so we should never be called for that
+	if f.cfdMode == common.CFDModeFlags.TargetCompare() {
+		panic("We should not be called for CFDMode==TargetCompare")
```
nit,
"HasFileChangedSinceLastSyncUsingLocalChecks() should not be called for CFDMode==TargetCompare"

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -36,16 +42,155 @@ type syncDestinationComparator struct {
+			return true, true
+		} else if so.lastChangeTime.After(f.lastSyncTime) {
+			// File Ctime changed only, only meta data changed.
+			return false, true
```
We need to distinguish between metadata only sync and data+metadata sync.
512.	            // 
513.	            // Else if ctime > LastSyncTime, then only metadata needs to be sync'ed. 
514.	            // If MetadataOnlySync is True then we distinguish between data 
515.	            // and metadata sync, else we always do data+metadata sync. 
516.	            // 
517.	            if MetadataOnlySync { 
518.	                return False, True 
519.	            } else { 
520.	                return True, True 
521.	            }

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -36,16 +42,155 @@ type syncDestinationComparator struct {
+	} else {
+		// This is the case when neither CtimeMtime or Ctime cfdMode set. So its target traverse and if we reach here
+		// means these entries for new files.
+		return true, true
```
we already checked and panic'ed for TargetCompare above.
Here we should add the same panic that says unsupported CFDmode and prints it

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -36,16 +42,155 @@ type syncDestinationComparator struct {
+			storedObject := folderMap.indexMap[file]
+			size += storedObjectSize(storedObject)
+			delete(folderMap.indexMap, file)
+			metaDataChange, dataChange := f.HasFileChangedSinceLastSyncUsingLocalChecks(storedObject, storedObject.relativePath)
```
the return value order is swapped, it should be
dataChange, metaDataChange

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -36,16 +42,155 @@ type syncDestinationComparator struct {
+			if dataChange {
+				f.copyTransferScheduler(storedObject)
+			} else if metaDataChange {
+				// TODO: Add calls to just update meta data of file.
```
let;s add a panic("Not yet implemented") to catch till we implement

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -36,16 +42,155 @@ type syncDestinationComparator struct {
+		}
+
+	}
+	size = -size
```
nit, remove this and use -size in the next line

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,191 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
 }
 
+//
+// Given both the local and target attributes of the file, find out if we need to copy data, metadata, both, none.
```
Given both the local and target attributes of the file, find out if we need to copy data+metadata, only metadata, or none.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,191 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+			// We would have set target object’s ctime equal to source object’s ctime when we last sync’ed the object, so if it’s not equal now,
+			// it means the source object’s metadata was updated since last sync.
+			//
+			if so.lastChangeTime.After(to.lastChangeTime) {
```
equality comparison not After

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,191 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+//
+// Unlike HasFileChangedSinceLastSyncUsingLocalChecks(), it needs the target attributes also, which means it can only
+// be called for cases where we enumerate the target.
+//
```
// Return: (dataChanged, metadataChanged)

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,191 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+					return true, true
+				}
+			}
+		} else if so.lastChangeTime.Sub(time.Time{}) != (to.lastChangeTime.Sub(time.Time{}) - time.Duration(time.Second*time.Duration(f.TargetCtimeSkew))) {
```
this is more readable? Note the minus sign

so.lastChangeTime.Sub(to.lastChangeTime) > -(time.Duration(f.TargetCtimeSkew) * time.Second)

nitin-deamon replied:
Please check the new change.

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,90 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+			}
+		}
+
 		for marker := (azblob.Marker{}); marker.NotDone(); {
```
comment above this for loop

		//
	    // Enumerate direct children of currentDirPath (aka non-recursively).
	    // Each child object we will call enqueueOutput() which will cause it to be sent to
	    // processIfNecessary() for processing where we will perform the "do we need to sync it"
	    // check. After currentDirPath is fully enumerated and all children queued for processing
	    // by processIfNecessary(), we also enqueue a special StoredObject to trigger the "end of
	    // directory" processing by processIfNecessary(). This StoredObject is marked by isFolderEndMarker
	    // flag. This indicates that we are done processing target objects. This could mean one of
	    // the following:
	    //
	    // 1. currentDirPath was found "not changed" and hence we didn't need to perform target
	    //    enumeration. In this case we need to go over all StoredObjects belonging to currentDirPath
	    //    in sourceFolderIndex and see if they are modified after last sync and transfer if needed.
	    // 2. currentDirPath was found "changed" and hence we enumerated the target and we are done
	    //    processing all enumerated children, so what's left now is new files/subdirs created in source
	    //    after lastSyncTime and we need to transfer them all. This is marked by setting FinalizeAll
	    //    to true.
	    // 3. currentDirPath was not found in the target. In this case we need to create the directory
	    //    in the target and transfer all its direct children. See Note below.
	    //
	    // Above is performed by the FinalizeTargetDirectory() method.
	    //
		// Note: For the case of sync target traverser currentDirPath may not exist in the target
		//       (since it was picked from tqueue which means it exists at source but not necessarily
		//       at the target). In that case the following enumeration will not yield anything, note
		//       that it'll not fail but simply result in "nothing". In that case we will go to
		//       FinalizeDirectory and create a "end of folder" marker to be sent to processIfNecessary().
		//       processIfNecessary() will perform the target directory creation.
		//

nitin-deamon replied:
Not added this comment as i like to discuss

linuxsmiths reviewed (commented):
Pushing ongoing changes.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,191 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+			}
+
+			lcFolderName = path.Join(lcFolderName, lcFileName)
+			foldermap, folderPresent = f.sourceFolderIndex.folderMap[lcFolderName]
```
just say

foldermap, folderPresent = f.sourceFolderIndex.folderMap[lcRelativePath]

nitin-deamon replied:
For root folder lcRelativePath will be "" and we want path to be ".". Thats why path join as lcFolderName and lcFileName in case of root folder will be "." and "." . path join give us "."

linuxsmiths replied:
ok, then this is something subtle that needs to be explained in the comment.
Let's add the following comment

            //
            // Note: We join lcFolderName and lcFileName to index into folderMap[] instead of directly indexing
            //       using lcRelativePath. This is because for the root directory, lcRelativePath will be "" but
            //       the SourceTraverser would have stored it's properties in folderMap["."], path.Join() gets us
            //       "." for root directory.
            //

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,191 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+
+			lcFolderName = path.Join(lcFolderName, lcFileName)
+			foldermap, folderPresent = f.sourceFolderIndex.folderMap[lcFolderName]
+			if folderPresent {
```
When would we not have folderPresent true?
Even for empty source directory or for the case where we have processed and delete all children, we must still have entry for "."
f.e. for an empty dir "dir1/dir2", we will have one entry

folderMap["dir1/dir2"]["."].

Right?

Then we should panic if we don't have it set

And this is the place we need to log "Finalizing directory %s (FinalizeAll=%v)"

nitin-deamon replied:
No Tomar , ["dir1/dir2"]["."] will be removed by getStoredObject, so if all file present on source and target same, folder will be deleted. Like in case if someone create the file and delete it  before next sync. Ctime changed and we do target traverse.

linuxsmiths replied:
As discussed let's delete folderMap[folder] only from FinalizeTargetDirectory since we finalize a directory when we are done processing all its children, so that's the right time to delete it from folderMap.

nitin-deamon committed:
Removed the code for isVirtualFolder.

- In latest azcopy code directorystubs are set as folder only.
So we don't need isVirtualFolder boolean.
- Some minor fixes which cause folder to be copied again in target traverser.

linuxsmiths commented on cmd/syncIndexer.go:
```diff
@@ -177,6 +178,66 @@ func (i *folderIndexer) store(storedObject StoredObject) (err error) {
+//
+func (i *folderIndexer) getStoredObject(relativePath string) StoredObject {
+	var lcRelativePath string
+	if i.isDestinationCaseInsensitive {
```
nit, leave one blank line after variable decl block and the code.

linuxsmiths commented on cmd/syncIndexer.go:
```diff
@@ -177,6 +178,66 @@ func (i *folderIndexer) store(storedObject StoredObject) (err error) {
+	} else {
+		lcRelativePath = relativePath
+	}
+	lcFolderName := filepath.Dir(lcRelativePath)
+	lcFileName := filepath.Base(lcRelativePath)
+	lcFolderName = path.Join(lcFolderName, lcFileName)
```
any reason for doing this, instead of just using lcRelativePath directly in L226 and below.
Technically we want to lookup foldermap["dir1/dir2"]["."], better to keep the code simple to read.
I think we need to do similar simplification in store() method of source traverser.

also leave a blank line before i.lock.Lock() so it's better visible.

nitin-deamon replied:
for root folder we want ["."]["."]

linuxsmiths replied:
ok, such subtle details are best documented for the benefit of the reader.
add this comment at L223

    //
    // Instead of directly indexing folderMap with lcRelativePath we do the join(Dir(), Base()) trick to
    // correctly support the root folder case where lcRelativePath is the empty string "", we don't
    // want to index folderMap[""] since that doesn't exist, instead we want to index root folder's
    // properties as folderMap["."]["."], since that's where folderIndexer.store() stores them.
    //

linuxsmiths commented on cmd/syncIndexer.go:
```diff
@@ -177,6 +178,66 @@ func (i *folderIndexer) store(storedObject StoredObject) (err error) {
+			delete(folderMap.indexMap, ".")
+			size := storedObjectSize(so)
+			size = -size
+			atomic.AddInt64(&i.totalSize, size)
```
nit, simply
atomic.AddInt64(&i.totalSize, -(storedObjectSize(so)))

also let's panic if i.totalSize becomes -ve

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,90 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+		// This code for sync operation and when its target traverser.
+		if t.isSync && !t.isSource {
+			so := t.indexerMap.getStoredObject(currentDirPath)
+			currentDirPath = so.relativePath
```
it's not clear why we are re-setting currentDirPath  to so.relativePath.
Instead we should panic if both are not same.
If you think currentDirPath  may not be canonical (f.e. it may have  slash at the end), then use filepath.Clean() to canonicalize the path.

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,90 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+
 func (t *blobTraverser) parallelList(containerURL azblob.ContainerURL, containerName string, searchPrefix string,
 	extraSearchPrefix string, preprocessor objectMorpher, processor objectProcessor, filters []ObjectFilter) error {
 	// Define how to enumerate its contents
 	// This func must be thread safe/goroutine safe
```
update comment

    //
    // Define how to enumerate its contents
    // This func must be thread safe/goroutine safe            
    //
    // For the sync case, this is called for each directory added to tqueue by the source traverser.
    // Note that since source traverser will add directories present in the source, it may or may not
    // be present in the target. If a directory is not present in the target, that's a case of newly
    // created directory at the source which must be created in the target.
    // It should enumerate only the direct children of the directory and create a storedObject for each
    // and queue them for processing by processIfNecessary.
    // Apart from queueing storedObject for each direct children, it'll also queue a special storedObject
    // for "finalize" processing. This will be queued *after* all the children.
    //

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -300,7 +382,9 @@ func (t *blobTraverser) parallelList(containerURL azblob.ContainerURL, container
 
 				storedObject := t.createStoredObjectForBlob(preprocessor, blobInfo, strings.TrimPrefix(blobInfo.Name, searchPrefix), containerName)
-
+				extendedProp, _ := common.ReadStatFromMetadata(blobInfo.Metadata, *blobInfo.Properties.ContentLength)
```
since we are continuing at L382 for directory stubs, how are we able to get properties correctly for folders.

nitin-deamon replied:
Yes Tomar.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,192 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+	var present bool
+	var sourceObjectInMap StoredObject
+
+	if f.sourceFolderIndex.isDestinationCaseInsensitive {
```
comment above this

    //
    // We come here with 'destinationObject' when Target traverser (parallelList() for Target=Blob)
    // enumerates a directory from tqueue and creates and enqueues a destinationObject for each
    // enumerated children.
    // Target traverser also creates and enqueues a special destinationObject (with isFolderEndMarker==true)
    // after queueing destinationObjects for all direct children of the directory.
    // This special destinationObject is guaranteed to be queued *after* all the other children objects for
    // the directory (if any).
    // See below for more details on the special isFolderEndMarker object.
    //

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,192 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
-			err := f.copyTransferScheduler(sourceObjectInMap)
-			if err != nil {
-				return err
+	lcFolderName = filepath.Dir(lcRelativePath)
```
comment above this

    //
    // parent dir name, will be "." for files/dirs at the copy root.
    //
    // So, for lcRelativePath == "dir1/dir2",
    // lcFolderName -> "dir1"
    // lcFileName   -> "dir2"
    //
    // and for "dir1" (dir1 present at copy root),
    // lcFolderName -> "."
    // lcFileName   -> "dir1"
    //
    // Q: When will folderPresent be false below?
    // A: Since we process every directory independently, if "dir1" is fully processed before "dir1/dir2",
    //    when processIfNecessary() is called for "dir1/dir2" with isFolderEndMarker==true, foldermap["dir1"]
    //    won't be found as it would have been deleted after "dir1" was fully processed, though we MUST have
    //    foldermap["dir1/dir2"].
    //    Also note that folderPresent MUST never be false when isFolderEndMarker==false, since target traverser
    //    only ever enumerates directories added to tqueue by the source traverser and all such directories
    //    MUST be added to folderMap by the source traverser.
    //

we can panic/assert for the above.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,192 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+			// 1. Dir not present on target side, then we need to create directory and files.
+			// 2. Dir present and some files are added.
+			// 3. Dir present and some files are modified.
+			// For each case we need to process left over files in ObjectIndexerMap.
```
consider the following elaborate comment

            //
            // We will see this special "end of folder" marker *after* we have seen all direct children of
            // that directory (if any). We have the following cases:
            // 1. destinationObject.relativePath directory does not exist in the target. In this case
            //    there won't be any children of destinationObject.relativePath queued for us and we need
            //    to copy the directory recursively to the target.
            // 2. destinationObject.relativePath directory exists in the target and the source directory
            //    has not "changed" since last sync. In this case the Target traverser would have skipped
            //    enumeration of destinationObject.relativePath and hence there won't be any children queued
            //    for us. We need to go over all its children added in sourceFolderIndex and test them for
            //    "newness" and copy them if they are found modified after lastsync.
            // 2. destinationObject.relativePath directory exists in the target and the source directory
            //    has "changed" since last sync. In this case Target traverser would have enumerated the
            //    directory and queued all enumerated children for us. After all the children it would have
            //    queued this special isFolderEndMarker object. So when we process the isFolderEndMarker object,
            //    we would have processed all children and hence whatever is left in sourceFolderIndex for
            //    that directory are only the files newly created in the source since last sync. We need to
            //    copy all of them to the target.
            //

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,192 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+						"\n If folder is empty and don't exist then will be created as parent enumeration done. Otherwise will be created by its child enumeration."+
+						"\n Its metadata will be set by parent only as it has the required info.", lcRelativePath)
+				}
+			}
```
I think we can omit this. This is more confusing and not adding value.
I've explained in the comment above why folderPresent  may or may not be false.
Let's say processIfNecessary() is called for endFolderMarker for "dir1/dir2". 
Whether we find folderMap for "dir1" depends on whether "dir1" was processed before "dir1/dir2", and since each folder is indepdently processed it could very well have been processed or maybe not, so there's not much point in the following print and check.
Remind me to discuss this.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,192 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+			}
+			return nil
+		} else {
+			if folderPresent {
```
as commented above, for destinationObject.isFolderEndMarker==false, folderPresent must never be false.

nitin-deamon replied:
No tomar, suppose dir1 has dir2, dir3, dir4 and other files. Same on target. Now we delete the dir4 from source.
We enumerate target it gives dir2, dir3, dir4 and files. Suppose dir4 came as last entry. In that case dir1 foldermap will be deleted. As no entry left.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,192 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+
+	// File case.
+	// Parent Folder present, we need to check if file exists or not.
+	if folderPresent {
```
again, we should never have the case that folderPresent  is false. we can panic on this,

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,192 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+	if folderPresent {
+		sourceObjectInMap, present = foldermap.indexMap[lcFileName]
+
+		// if the destinationObject is present at source and stale, we transfer the up-to-date version from source
+		if present && (sourceObjectInMap.entityType == destinationObject.entityType) {
+			// Sanity check.
+			if sourceObjectInMap.relativePath != destinationObject.relativePath {
+				panic("Relative Path not matched")
+			}
+
+			dataChanged, metadataChanged := f.HasFileChangedSinceLastSyncUsingTargetCompare(destinationObject, sourceObjectInMap)
+			if f.disableComparison || dataChanged {
+				err := f.copyTransferScheduler(sourceObjectInMap)
+				if err != nil {
+					return err
+				}
+			} else if metadataChanged {
+				// TODO: Need to add call to just update the metadata only.
+			}
+			size := storedObjectSize(sourceObjectInMap)
+			size = -size
+			atomic.AddInt64(&f.sourceFolderIndex.totalSize, size)
+			delete(f.sourceFolderIndex.folderMap[lcFolderName].indexMap, lcFileName)
+			return nil
+		}
 	}
 
+	// Parent folder not present this may happen when all the entries for that folder in folder map processed.
+	// In that case the foldemap will be deleted, so we know this file not present in source.
```
again, we need this code from doc

                //
                // tobject is a file.
                //          
                // If not present in source, it means the object was deleted
                // from the source since last sync.
                //
                if !IsPresent(ObjectIndexer[dirName][baseName]) {
                    //
                    // Delete the target file unless KeepDeletedFiles is true.
                    //
                    if (!KeepDeletedFiles) {
                       //
	             // For directories DeleteFromTarget will do recursive
	             // deletion.
	             //
	             DeleteFromTarget(tobject)
	          }
	      } else {
	          //
	          // Object with same pathname exists at the source.
	          // Get the storedObject from the ObjectIndexer map,
	          // and delete from ObjectIndexer map.
	          // We delete all processed objects from ObjectIndexer
	          // map so that at the end whatever is left in the map
	          // are new objects created at the source since last
	          // sync and we can then copy them to target.
	          //
	          storedObject = ObjectIndexer[dirName][baseName]
	          delete(ObjectIndexer[dirName][baseName])
	          if (IsFile(storedObject.Attributes)) {
	              //
	              // Target is a regular file and so is the source.
	              //    
	              //
	              // Find out whether file's data or metadata or both
	              // have changed since last sync.
	              // Note that since a "changed directory" can be a renamed
	              // directory, we need to compare each of its children with
	              // their counterpart in the target to find out if they
	              // need to be synced.
	              //
	              dataChanged, metadataChanged =
	           HasFileChangedSinceLastSyncUsingTargetCompare(tobject, storedObject)
	         
	              if dataChanged {
	                  CopyFileToTarget(storedObject)
	              } else if metadataChanged {
	                  CopyFileMetadataToTarget(storedObject)
	              }                    
	          } else if (!IsDirectory(storedObject.Attributes)) {
	              //
	              // Source is a regular file but target is some other
	              // non-directory object like symlink, device file, etc.
	              //
	              DeleteFromTarget(tobject)
	              CopyFileToTarget(storedObject)
	          } else {
	               //
	               // Source is a regular file but target is a directory
	               // (with the same name).
	               //
	               DeleteFromTarget(tobject)
	               CopyFileToTarget(storedObject)
	          }
	      }

let's discuss

linuxsmiths reviewed (commented):
Completely reviewed parallelList() and processIfNecessary().
Pls look at the comments, address and let me know if need to discuss.

nitin-deamon commented on cmd/syncComparator.go:
```diff
@@ -54,30 +199,192 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+		} else {
+			if folderPresent {
+				sourceObjectInMap, present = foldermap.indexMap[lcFileName]
+				// Lets check if entry in source is folder and entry on destination should be folder.
+				// Otherwise folder in source changed to file.
+				if present && (destinationObject.entityType == sourceObjectInMap.entityType) {
+					delete(foldermap.indexMap, lcFileName)
+					size := storedObjectSize(sourceObjectInMap)
+					size = -size
+					atomic.AddInt64(&f.sourceFolderIndex.totalSize, size)
+					if sourceObjectInMap.relativePath != destinationObject.relativePath {
+						panic("Relative Path not matched")
+					}
+					delete(f.sourceFolderIndex.folderMap[lcFolderName].indexMap, lcFileName)
+
+					// In case of folder if ctime/mtime changed, lets create the folder.
+					dataChanged, metadataChanged := f.HasFileChangedSinceLastSyncUsingTargetCompare(destinationObject, sourceObjectInMap)
+					if f.disableComparison || dataChanged || metadataChanged {
+						err := f.copyTransferScheduler(sourceObjectInMap)
+						if err != nil {
+							return err
+						}
+					}
+					return nil
+				}
+			}
+			// We detect folder not present on source, now we need to delete the folder and files underneath.
+			// Other case will be folder at source changed to file, so we need to delete the folder in target.
+			// TODO: Need to add call to delete the folder.
+			_ = f.destinationCleaner(destinationObject)
```
Lets discuss it.

nitin-deamon committed:
Addressed the review comments -2

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +202,217 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+		} else {
+			if folderPresent {
+				sourceObjectInMap, present = foldermap.indexMap[lcFileName]
+				// Lets check if entry in source is folder and entry on destination should be folder.
+				// Otherwise folder in source changed to file.
```
//
                // Target also has this object and is of the same type (directory) as the source?
                // If yes, do nothing right now. When we process this directory from tqueue and finalize it,
                // that time we will sync the directory properties if needed.
                // If either target does not have object with the same name or object type is different, we
                // will need to delete it. This will be done by the f.destinationCleaner() call below.
                // The new directory will be created in FinalizeTargetDirectory() when we go over the folderIndexer
                // map entries (note that we don't remove it from the folderIndexer here).
                //

linuxsmiths reviewed (commented):
still looking.. waiting for your test results

linuxsmiths commented on cmd/sync.go:
```diff
@@ -90,15 +90,31 @@ type rawSyncCmdArgs struct {
+	// in the past. 60 seconds is a good value for skew adjustment. A larger skew value could cause more data to be sync'ed while a smaller skew value may
+	// cause us to miss some changed files, latter is obviously not desirable. If we are not sure what skew value to use, it's best to create a temp file
+	// on the source and compare its ctime with the nodes time and use that as a baseline.
+	//
```
for uniformity, let's move this description also to cookedSyncCmdArgs and put a ref to the comment here like other variables.
Here we can just say
// This is the time of last sync in ISO8601 format. For more information, please refer to cookedSyncCmdArgs.

linuxsmiths commented on cmd/sync.go:
```diff
@@ -115,6 +131,19 @@ func (raw *rawSyncCmdArgs) parsePatterns(pattern string) (cookedPatterns []strin
+	} else if cfdMode == strings.ToLower(common.CFDModeFlags.CtimeMtime().String()) {
+		return common.CFDModeFlags.CtimeMtime()
+	} else {
+		return common.CFDModeFlags.TargetCompare()
```
let's not allow any arbitrary string but instead let's fail if user provides an invalid string.
To support the case where user does not provide the option and we want to default to TargetCompare, check for empty string explicitly.

linuxsmiths commented on cmd/sync.go:
```diff
@@ -250,10 +279,24 @@ func (raw *rawSyncCmdArgs) cook() (cookedSyncCmdArgs, error) {
 	}
 
+	// Process the change directory mode for Sync operation.
+	cooked.cfdMode = raw.parseCFDMode()
```
fix the comment

// Process the change file detection mode for Sync operation.

and return error from parseCFDMode() and bail out on error, if user provides an invalid string as option.

linuxsmiths commented on cmd/sync.go:
```diff
@@ -250,10 +279,24 @@ func (raw *rawSyncCmdArgs) cook() (cookedSyncCmdArgs, error) {
+			return cooked, err
+		}
+	} else {
+		cooked.lastSyncTime = time.Time{}
```
if CFDMode is either of CtimeMtime or Ctime, and user has not provided lastSyncTime, we should bail out.
This is part of arg validation that we should do in this function that converts raw args to cooked args

linuxsmiths commented on cmd/sync.go:
```diff
@@ -851,6 +917,20 @@ func init() {
 	syncCmd.PersistentFlags().BoolVar(&raw.preservePermissions, PreservePermissionsFlag, false, "False by default. Preserves ACLs between aware resources (Windows and Azure Files, or ADLS Gen 2 to ADLS Gen 2). For Hierarchical Namespace accounts, you will need a container SAS or OAuth token with Modify Ownership and Modify Permissions permissions. For downloads, you will also need the --backup flag  ...
 
+	// Changed file detection mode.
+	syncCmd.PersistentFlags().StringVar(&raw.cfdMode, "cfd-mode", "TargetCompare", " cfd-mode is Change File Detection Mode. It has valid values TargetCompare(default), Ctime, CtimeMtime, ArchiveBit."+
```
let's remove ArchiveBit since we don't support it

linuxsmiths commented on cmd/sync.go:
```diff
@@ -851,6 +917,20 @@ func init() {
+	syncCmd.PersistentFlags().BoolVar(&raw.metaDataOnlySync, "metadata-only-sync", false, "Optimization to transfer only metaData in case of metadata change only.")
+
+	// Time from which file change detection done.
+	syncCmd.PersistentFlags().StringVar(&raw.lastSyncTime, "last-sync-time", "", "Include files modified after lastSyncTime.")
```
We need to tell the user what format the string has to be in, i.e. ISO8601 datetime format, and also give a sample string for ease of understanding.
is there already an argument that takes time, what is the descr for that.
let's discuss this.

linuxsmiths commented on cmd/syncIndexer.go:
```diff
@@ -177,6 +178,74 @@ func (i *folderIndexer) store(storedObject StoredObject) (err error) {
+			atomic.AddInt64(&i.totalSize, -storedObjectSize(so))
+			if atomic.LoadInt64(&i.totalSize) < 0 {
+				panic("Total Size is negative.")
+			}
```
let's move this to FinalizeDirectory, so that we only delete folderMap["dir"]["."] in FinalizeDirectory which is the last function to be called for a directory.

linuxsmiths commented on cmd/syncIndexer.go:
```diff
@@ -177,6 +178,74 @@ func (i *folderIndexer) store(storedObject StoredObject) (err error) {
+			return so
+		}
+	}
+	panic(fmt.Sprintf("Stored Object for relative path[%s] not found", lcFolderName))
```
let's log both lcRelativePath and lcFolderName, just in case some of the Dir(), Base() and Join() above resulted in some issue.

linuxsmiths commented on cmd/zc_enumerator.go:
```diff
@@ -88,11 +89,9 @@ type StoredObject struct {
-	isVirtualFolder bool
+	isFolderEndMarker bool
+
+	isFinalizeAll bool
```
//
    // For every folder processed by TargetTraverser it queues a special StoredObject (to be processed by
    // processIfNecessary()) to mark the "end of directory". This causes the FinalizeTargetDirectory() to be
    // called which does the "end of directory" processing like copying remaining files, etc.
    // isFolderEndMarker marks this special "end of folder" StoredObject.
    // isFinalizeAll is passed as an argument to FinalizeTargetDirectory() and it conveys whether TargetTraverser
    // enumerated and processed contents of the directory or it did not (since the directory was considered
    // "not changed since last sync".
    //

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,144 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+			// files with same names as in the target could be entirely different files. This forces us to enumerate the target directory if the source
+			// directory is seen to have changed, since we don’t know if it was renamed, in which case we must enumerate the target directory.
+			//
+			if !t.HasDirectoryChangedSinceLastSync(so, containerURL) {
```
It's more logical to pass currentDirPath  string to HasDirectoryChangedSinceLastSync and do the getStoredObject stuff inside it. Also as commented earlier, pls remove containerURL argument, it's not required.

nitin-deamon replied:
I added containerURL to get target properties. Now we removed that code, yes we don't need this.

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -236,14 +239,144 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+
 func (t *blobTraverser) parallelList(containerURL azblob.ContainerURL, containerName string, searchPrefix string,
 	extraSearchPrefix string, preprocessor objectMorpher, processor objectProcessor, filters []ObjectFilter) error {
+	// Find whether call for target Traverser or not.
```
comment can say

    //
    // If we are scanning the target in sync mode (aka TargetTraverser) we need to do some special handling.
    // Set the following boolean to be used later in this function.
    //

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -253,26 +386,31 @@ func (t *blobTraverser) parallelList(containerURL azblob.ContainerURL, container
 						// try to get properties on the directory itself, since it's not listed in BlobItems
 						fblobURL := containerURL.NewBlobURL(strings.TrimSuffix(currentDirPath+virtualDir.Name, common.AZCOPY_PATH_SEPARATOR_STRING))
 						resp, err := fblobURL.GetProperties(t.ctx, azblob.BlobAccessConditions{}, azblob.ClientProvidedKeyOptions{})
+						// TODO: Need to debug why its giving error here.
```
is it still giving error?

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -253,26 +386,31 @@ func (t *blobTraverser) parallelList(containerURL azblob.ContainerURL, container
 								getObjectNameOnly(strings.TrimSuffix(currentDirPath+virtualDir.Name, common.AZCOPY_PATH_SEPARATOR_STRING)),
 								strings.TrimSuffix(currentDirPath+virtualDir.Name, common.AZCOPY_PATH_SEPARATOR_STRING),
-								common.EEntityType.File(), // folder stubs are treated like files in in the serial lister as well
+								common.EEntityType.Folder(), // folder stubs are treated like files in in the serial lister as well
```
the comment doesn't make sense now since we are already passing Folder()

linuxsmiths commented on cmd/zc_traverser_blob.go:
```diff
@@ -334,6 +474,23 @@ func (t *blobTraverser) parallelList(containerURL azblob.ContainerURL, container
+				entityType:        common.EEntityType.Folder(),
+				lastModifiedTime:  time.Time{},
+				size:              0,
+				ContainerName:     containerName,
```
if we don't use them we can omit them so that they default to their default values and it helps readability also.

linuxsmiths commented on common/fe-ste-models.go:
```diff
@@ -1322,6 +1322,24 @@ func (e EntityType) String() string {
+
+func (CFDMode) Ctime() CFDMode { return CFDMode(2) }
+
+func (CFDMode) CtimeMtime() CFDMode { return CFDMode(3) }
```
let's follow the same order in all places

TargetCompare, CtimeMtime, Ctime

linuxsmiths commented on common/parallel/TreeCrawler.go:
```diff
@@ -291,7 +291,7 @@ func (c *crawler) processOneDirectoryWithAutoPacer(ctx context.Context, workerIn
 			// 2. mayHaveMoreDirs is true, which means we are not yet done reading from tqueue and hence more new directories may be added by source traverser.
 			//
-			for len(c.unstartedDirs) == 0 && c.mayHaveMoreDirs && ctx.Err() == nil {
+			for len(c.unstartedDirs) == 0 && (c.mayHaveMoreDirs || c.dirInProgressCount > 0) && ctx.Err() == nil {
```
we discussed that for target traverser we don't need to check dirInProgressCount since we are not depending on BFS enumeration to add more dirs, instead we only depend on whatever is added to tqeue.
Why did we have to add this?
let's discuss. Also if needed we should update the comment explainign why

nitin-deamon replied:
lets discuss this, i am hitting panic in case, still directories in progress, But source done. We hit panic down.

linuxsmiths commented on common/sysinfo_linux.go:
```diff
@@ -0,0 +1,154 @@
+		if strings.HasPrefix(scanner.Text(), "MemAvailable") {
+			var multipler int64
+			var result int
+			re := regexp.MustCompile(` +`)
+			value := strings.Split(string(re.ReplaceAll([]byte(scanner.Text()), []byte(" "))), " ")[1]
+			multiplerStr := strings.Split(string(re.ReplaceAll([]byte(scanner.Text()), []byte(" "))), " ")[2]
```
we can avoid regexp and use strings.Fields() which returns tokens separated by whitespace.
also typo multipler ->multiplier
also for sanity check for 

tokens := strings.Fields(scanner.Text())

if len(tokens) != 3 {
// errr
}

if MemAvailable is not there, we can do the following
"echo 3 > /proc/sys/vm/drop_caches" and then check for MemFree

linuxsmiths commented on common/sysinfo_linux.go:
```diff
@@ -0,0 +1,154 @@
+		err := fmt.Errorf("GetMemInfo failed with error: %v", err)
+		return 0, err
+	}
+
```
this should be a failure return since we couldn't get valid available memory

nitin-deamon replied:
Yes, but we need add the memfree code here .

linuxsmiths commented on common/sysinfo_linux.go:
```diff
@@ -0,0 +1,154 @@
+// MemAvailable field added after 3.14+ kernel only. So for kernel version before that we need to give rough estimate.
+//
+// TODO: Estimating MemAvailable on kernel before 3.14.
+func GetMemInfo() (int64, error) {
```
rename to GetMemAvailable

linuxsmiths commented on common/sysinfo_linux.go:
```diff
@@ -0,0 +1,154 @@
+		// dirfd is a null pointer, because we should only ever be passing relative paths here, and directories will be passed via transferInfo.Source.
+		// AT_SYMLINK_NOFOLLOW is not used, because we automagically resolve symlinks. TODO: Add option to not follow symlinks, and use AT_SYMLINK_NOFOLLOW when resolving is disabled.
+		err := unix.Statx(0, fileName,
+			unix.AT_STATX_SYNC_AS_STAT,
```
I think we should use AT_SYMLINK_NOFOLLOW since we are a low level function and caller if he wants can resolve symlink before calling GetExtendedProperties().
let's discuss this.

linuxsmiths reviewed (commented):
Reviewed everything except syncComparator.go.
Will take one last look at syncComparator.go after you make changes discussed till now.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -36,16 +42,158 @@ type syncDestinationComparator struct {
+				panic("Not yet implemented")
+			}
+		}
+	}
```
after the for loop when we are done creating any new files are directories. we must update the directory metadata, since we are done processing all objects inside this directory. This is also when we should delete the folder map for this directory as by definition finalize must be called in the end after all objects in the directory are processed.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -36,16 +42,158 @@ type syncDestinationComparator struct {
+	// 2. Does only metadata need to be sync'ed?
+	// 3. Object is not changed since last sync, no need to update.
+	//
+
```
we should count the number of directories finalized and also the number of directories we have dequeued from tqueue. These should match and must be equal to the total number of directories scanned by the source.
Such checks are important for resilience to make sure we don't miss anything due to some bug.

nitin-deamon replied:
Lets do it in next iteration.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -36,16 +42,158 @@ type syncDestinationComparator struct {
+			delete(folderMap.indexMap, file)
+			dataChange, metaDataChange := f.HasFileChangedSinceLastSyncUsingLocalChecks(storedObject, storedObject.relativePath)
+			if dataChange {
+				f.copyTransferScheduler(storedObject)
```
if storedObject is a directory will this correct cause the directory to be created.
For flat blobs directory creation may not do anything, but at least the stub has to be created.

nitin-deamon replied:
It will create the stub and its properties.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -36,16 +42,158 @@ type syncDestinationComparator struct {
+			storedObject := folderMap.indexMap[file]
+			size += storedObjectSize(storedObject)
+			delete(folderMap.indexMap, file)
+			f.copyTransferScheduler(storedObject)
```
For the case of newly created directory and files inside it, we will come to FinalizeTargetDirectory() (possibly) with a non-existent parent directory. Will this also create all the non-existent ancestor directories and their stubs?

nitin-deamon replied:
Yes Tomar.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +202,217 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+	defer func() {
+		if folderPresent {
+			if len(f.sourceFolderIndex.folderMap[lcFolderName].indexMap) == 0 {
+				size := int64(unsafe.Sizeof(objectIndexer{}))
+				atomic.AddInt64(&f.sourceFolderIndex.totalSize, -size)
+				delete(f.sourceFolderIndex.folderMap, lcFolderName)
 			}
 		}
-	} else {
-		// purposefully ignore the error from destinationCleaner
-		// it's a tolerable error, since it just means some extra destination object might hang around a bit longer
-		_ = f.destinationCleaner(destinationObject)
+		f.sourceFolderIndex.lock.Unlock()
+	}()
```
as discussed let's delete folderMap[folder] from FinalizeTargetDirectory. This is not needed then.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +202,217 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+					size = -size
+					atomic.AddInt64(&f.sourceFolderIndex.totalSize, size)
+					if sourceObjectInMap.relativePath != destinationObject.relativePath {
+						panic("Relative Path not matched")
```
also log the two relative paths for better debuggability, if this happens

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -54,30 +202,217 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+				if err != nil {
+					return err
+				}
+			} else if metadataChanged {
+				// TODO: Need to add call to just update the metadata only.
+			}
+			size := storedObjectSize(sourceObjectInMap)
+			size = -size
+			atomic.AddInt64(&f.sourceFolderIndex.totalSize, size)
```
need to be careful we don't miss any of these.. I prefer adding panic in all unsupported places so that we know if it runs this code.
also, for uniformity use

atomic.AddInt64(&f.sourceFolderIndex.totalSize, -(storedObjectSize(sourceObjectInMap)))

in all places

linuxsmiths reviewed (commented):
submitting pending comments

nitin-deamon committed:
Addressed the review comments - 3

linuxsmiths commented on common/sysinfo_linux.go:
```diff
@@ -0,0 +1,163 @@
+
+	stdOut, err := cmd.StdoutPipe()
+	if err != nil {
+		err := fmt.Errorf("GetMemInfo failed with error: %v", err)
```
GetMemInfo  -> GetMemAvailable
here and L27,74

linuxsmiths commented on common/sysinfo_linux.go:
```diff
@@ -0,0 +1,163 @@
+			var multiplier int64
+			var result int
+			tokens := strings.Fields(scanner.Text())
+			if len(tokens) < 3 {
```
"len(tokens) != 3" is a more accurate check.

linuxsmiths commented on common/sysinfo_linux.go:
```diff
@@ -0,0 +1,163 @@
+				multiplier = 1024
+			} else {
+				// "/proc/meminfo" output always in kB only. If we are getting different string, something wrong.
+				err := fmt.Errorf("Meminfo value is not in kB")
```
log scanner.Text() for debuggability

nitin-deamon committed:
Few NitPiks left in previous commit.

linuxsmiths commented on cmd/zc_enumerator.go:
```diff
@@ -91,6 +91,12 @@ type StoredObject struct {
 
 	isFolderEndMarker bool
 
+	//
+	// For every folder processed by TargetTraverser it queues a special StoredObject (to be processed by processIfNecessary()) to mark the "end of directory".
+	// This causes the FinalizeTargetDirectory() to be called which does the "end of directory" processing like copying remaining files, etc. isFolderEndMarker
+	// marks this special "end of folder" StoredObject. isFinalizeAll is passed as an argument to FinalizeTargetDirectory() and it conveys whether TargetTraverser
+	// enumerated and processed contents of the directory or it did not (since the directory was considered "not changed since last sync".
+	//
```
this comment should come above isFolderEndMarker.
Put both the bools one after another (with no newline between them) and have this comment above the pair

// comment here <---
isFolderEndMarker bool
isFinalizeAll bool

linuxsmiths commented on cmd/syncIndexer.go:
```diff
@@ -231,19 +231,13 @@ func (i *folderIndexer) getStoredObject(relativePath string) StoredObject {
-			atomic.AddInt64(&i.totalSize, -storedObjectSize(so))
-			if atomic.LoadInt64(&i.totalSize) < 0 {
-				panic("Total Size is negative.")
+				panic(fmt.Sprintf("StoredObject for relative path[%s] and folderName[%s] is not of type folder", relativePath, lcFolderName))
```
sorry, github doesn't let me add comment in existing code, so adding it here
the comment above this function now should be updated, since we call it only for a specific purpose

        //
        // Given the relative path of a directory this returns the StoredObject corresponding to that directory.
        // To be precise it returns the StoredObject stored in folderMap["dir1/dir2"]["."] for directory named
        // "dir1/dir2."
        //
        // This MUST be called *only* by HasDirectoryChangedSinceLastSync() when it wants to find out if a directory
        // dequeued from 'tqueue' must be enumerated at the target. Such a StoredObject would have been added by the
        // source traverser for each directory it adds to tqueue.
        //

and add following coment at L224 to explain the path.join

                //
                // Note: We join lcFolderName and lcFileName to index into folderMap[] instead of directly indexing
                //       using lcRelativePath. This is because for the root directory, lcRelativePath will be "" but
                //       the SourceTraverser would have stored it's properties in folderMap["."], path.Join() gets
                //       us "." for root directory.
                //

nitin-deamon replied:
No worries Tomar, Thanks

linuxsmiths commented on cmd/sync.go:
```diff
@@ -131,16 +119,17 @@ func (raw *rawSyncCmdArgs) parsePatterns(pattern string) (cookedPatterns []strin
+		return common.CFDModeFlags.Ctime(), nil
 	} else {
-		return common.CFDModeFlags.TargetCompare()
+		err := fmt.Errorf("Invalid input of cfd-mode.")
```
("Invalid value for cfd-mode: %s", raw.cfdMode)

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -164,33 +165,83 @@ func (f *syncDestinationComparator) FinalizeTargetDirectory(folderMap *objectInd
+	// 5. [dir2][file2]
+	// For dir1 finalizer we update folder properties of dir1 and file1. On dir2 finalizer we do dir2 folder properties updation and file2.
+	// As finalizer should be called for each folder, and it make sense to update its properties after creating its child files.
+	//
```
pls consider this

	//
	// Note: Finalizer is called for *every* directory present in the source and it's guaranteed to be called
	//       after all children of the directory are processed. To be precise, when Finalizer is called for a
	//       directory we will need to process all remaining children of that directory before we return from
	//       the Finalizer. Put another way, once Finalizer for a directory returns we shouldn't be looking at
	//       any children of the directory and hence, f.e., for directory "dir1/dir2"" all folderMap entries
	//       with folderMap["dir1/dir2"] will be deleted before we return from the Finalizer.
	//
	// For *every* folder we update its properties in its finalizer.
	// f.e., for following directory structure present on source
	//
	// dir1/dir2
	// dir1/file1
	// dir2/file2
	//
	// the object indexer map entries will be following :-
	//
	// ["dir1"]["."]
	// ["dir1"]["dir2"]
	// ["dir1"]["file1"]
	// ["dir2"]["."]
	// ["dir2"]["file2"]
	// ["dir1/dir2"]["."]
	//
	// In dir1's Finalizer we do the following:
	//
	// 1. Create target file dir1/file1.
	// 2. Create target directory dir1/dir2.
	//    Note: As an optimization, or for some targets, we can skip this as component directories will be
	//          auto-created when a resident file is created, but note that this would mean that we won't
	//          create empty directories in the target.
	// 3. Update properties for target directory dir1.
	//
	// Note that we can safely update the dir1 properties as this is after the last chidren of dir1 has been
	// created.
	//

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -164,33 +165,83 @@ func (f *syncDestinationComparator) FinalizeTargetDirectory(folderMap *objectInd
 	// So, we don't do any ctime and mtime comparsion.
 	if finalizeAll {
 		for file := range folderMap.indexMap {
+			// Folder properties will be updated in the end. After all the files created.
```
//
			// ["dir1"]["."] is a special StoredObject for holding the properties for directory "dir1".
			// This need not be created, so skip it now. In the end, once we are done creating all the files/dirs
			// inside "dir1" we will update dir1's properties from ["dir1"]["."].
			//

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -164,33 +165,83 @@ func (f *syncDestinationComparator) FinalizeTargetDirectory(folderMap *objectInd
 
 	// For FinalizeAll flag true, we know target traverser done. Now the files left in ObjectIndexerMap are the new ones.
 	// So, we don't do any ctime and mtime comparsion.
 	if finalizeAll {
```
this code can be refactored as follows for less duplication. This is inline with doc too.

                for file := range folderMap.indexMap {
                    //
                    // ["dir1"]["."] is a special StoredObject for holding the properties for directory "dir1".
                    // This need not be created, so skip it now. In the end, once we are done creating all the
                    // files/dirs inside "dir1" we will update dir1's properties by querying from ["dir1"]["."].
                    //
                    if file == "." {
                        continue
                    }

                    storedObject := folderMap.indexMap[file]
                    size += storedObjectSize(storedObject)
                    delete(folderMap.indexMap, file)


                    // If finalizeAll==true we need to blindly copy *all* files/folders present in folderMap.indexMap.
                    dataChange, metaDataChange := true, true

                    // else, we need to find out if the file/folder has changed since last sync.
                    if (finalizeAll == false) {
                        dataChange, metaDataChange := f.HasFileChangedSinceLastSyncUsingLocalChecks(storedObject)
                    }

                    if dataChange {
                        // For folders, this should create a new empty folder if not present.
                        f.copyTransferScheduler(storedObject)
                    } else if metaDataChange {
                        // For folders, properties are sync'ed when the Finalizer for that folder is called.
                        if storedObject.entityType != common.EEntityType.Folder() {
                            panic("File properties to be sync'ed. Not yet implemented")
                        }
                    }
                }

linuxsmiths replied:
for now, to prevent the above  f.copyTransferScheduler(storedObject) call for a folder, you can add the following code above this

          // TODO: Remove this code.
          //       Till we add support for updating directory attributes, we prevent this directory
          //       creation, else this clashes with the directory creation at the end of FinalizeTargetDirectory().
          //       Once we add support for directory attributes updation, change that call to cause
          //       directory attributes updation and not directory creation and remove this code.
          //
          if storedObject.entityType == common.EEntityType.Folder() {
              dataChange = false
          }

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -155,7 +155,8 @@ func (f *syncDestinationComparator) HasFileChangedSinceLastSyncUsingLocalChecks(
 //
-func (f *syncDestinationComparator) FinalizeTargetDirectory(folderMap *objectIndexer, finalizeAll bool) {
+// This function will also update the folder metaData, if changed which is in case of finalizeAll true and delete the flderMap from the ObjectIndexerMap.
+func (f *syncDestinationComparator) FinalizeTargetDirectory(folderMap *objectIndexer, finalizeAll bool, lcFolderName string) {
```
can we just pass lcFolderName  and not folderMap?
we have the folderIndexed in syncDestinationComparator, right?
we can get to the foldermap from there?
That way the prototype also matches the doc and avoids passing args that can be avoided.

nitin-deamon replied:
sure

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -36,16 +42,209 @@ type syncDestinationComparator struct {
+	delete(folderMap.indexMap, ".")
+	if finalizeAll {
+		f.copyTransferScheduler(so)
+	}
```
write a comment above this

                //
                // XXX 
                // We actually want to update the directory properties and *not* creat a new directory.
                // Till we have support for updating the directory properties, we use this same call to
                // create the directory.
                // XXX
                //

nitin-deamon committed:
Addressed review comments -4

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -156,66 +156,91 @@ func (f *syncDestinationComparator) HasFileChangedSinceLastSyncUsingLocalChecks(
+func (f *syncDestinationComparator) FinalizeTargetDirectory(lcFolderName string, finalizeAll bool) {
 	var size int64
+
+	// No need to check folderPresent, as it's already checked by caller.
+	folderMap, _ := f.sourceFolderIndex.folderMap[lcFolderName]
+
```
at least let's panic if it's not found, accompanied by a comment why we are asserting

also, what happens if lcFolderName is "" i.e. it's called for root folder. I see that you are depending on the caller not calling with that.
Also, same holds true for the above assertion that you make. Instead, it would be better to do all this here.

I suggest do the following here

change arg lcFolderName to relativePath, indicating that it's a relative path of the directory.
then we do all the checks here, and remove these from the caller

        // Root folder is represented as "." in indexer map.
        if relativePath == "" {
            relativePath = "."
        }

        folderMap, folderPresent := f.sourceFolderIndex.folderMap[relativePath]

        if !folderPresent {
                panic(fmt.Sprintf("Folder with relativePath[%s] not present in ObjectIndexerMap", relativePath))
        }

        fmt.Printf("Finalizing directory %s (FinalizeAll=%v)\n", relativePath, finalizeAll)

nitin-deamon replied:
Agreed, i thought of the same. But my thought now this Finalizer getting long. Other reason for that, its single thread and nobody else delete the map except the processIfNecessary.

linuxsmiths commented on cmd/syncComparator.go:
```diff
@@ -316,14 +346,15 @@ func (f *syncDestinationComparator) processIfNecessary(destinationObject StoredO
 		}
 
 		fmt.Printf("Finalizing directory %s (FinalizeAll=%v)\n", lcFolderName, destinationObject.isFinalizeAll)
-		f.FinalizeTargetDirectory(foldermap, destinationObject.isFinalizeAll, lcFolderName)
```
see comment above.
remove all this

linuxsmiths commented on common/sysinfo_linux.go:
```diff
@@ -71,7 +71,7 @@ func GetMemAvailable() (int64, error) {
 
 	if err = cmd.Wait(); err != nil {
-		err := fmt.Errorf("GetMemInfo failed with error: %v", err)
+		err := fmt.Errorf("GetMemAvailable failed with error: %v", err)
```
update all other places in this func..
sorry, github won't let me comment there

nitin-deamon replied:
I should be sorry i missed this. Shame on me.

linuxsmiths commented on cmd/zc_traverser_local.go:
```diff
@@ -463,21 +464,27 @@ func (t *localTraverser) Traverse(preprocessor objectMorpher, processor objectPr
 				if t.incrementEnumerationCounter != nil {
 					t.incrementEnumerationCounter(entityType)
 				}
+				ctime := fileInfo.Sys().(*syscall.Stat_t).Ctim
```
Sys() may be null on Windows and/or may not be of type *syscall.Stat_t, so it doesn't look safe to do this in common code.

nitin-deamon replied:
I changed it to know the performance number. Intention is to use GetExtendedProperties which will be platform dependent code. Reverted back to original

nitin-deamon committed:
Addressed Review comments -5

nitin-deamon committed:
Addressed review comments -6

linuxsmiths reviewed (approved):
Thanks for patiently working on this, Nitin!
